mono/metadata/object.c

```diff
@@ -3955,36 +3955,46 @@ mono_runtime_get_main_args (void)
 {
 	MONO_REQ_GC_UNSAFE_MODE;
 	MonoError error;
-	MonoArray *result = mono_runtime_get_main_args_checked (&error);
+	MonoArrayHandle arg_array = mono_runtime_get_main_args_handle (&error);
+	MonoArray *result = MONO_HANDLE_RAW (arg_array);
 	mono_error_assert_ok (&error);
 	return result;
 }
 
+static MonoStringHandle
+string_from_main_arg (MonoDomain *domain, int idx, MonoError *error)
+{
+	HANDLE_FUNCTION_ENTER ();
+	error_init (error);
+	MonoStringHandle res = mono_string_new_handle (domain, main_args [idx], error);
+	HANDLE_FUNCTION_RETURN_VAL (res);
+}
+
 /**
- * mono_runtime_get_main_args_checked:
+ * mono_runtime_get_main_args_handle:
  * \param error set on error
  * \returns a \c MonoArray with the arguments passed to the main
  * program. On failure returns NULL and sets \p error.
  */
-MonoArray*
-mono_runtime_get_main_args_checked (MonoError *error)
+MonoArrayHandle
+mono_runtime_get_main_args_handle (MonoError *error)
 {
-	MonoArray *res;
+	HANDLE_FUNCTION_ENTER ();
+	MonoArrayHandle res;
 	int i;
 	MonoDomain *domain = mono_domain_get ();
-
 	error_init (error);
 
-	res = (MonoArray*)mono_array_new_checked (domain, mono_defaults.string_class, num_main_args, error);
-	return_val_if_nok (error, NULL);
+	res = mono_array_new_handle(domain, mono_defaults.string_class, num_main_args, error);
+	mono_error_assert_ok (error);
```

Comment:
There's legitimate reasons for `mono_array_new_handle` to fail: don't assert do `if (!is_ok (error)) goto leave`. since this function uses `HANDLE_FUNCTION_ENTER`/`HANDLE_FUNCTION_RETURN_*`, it needs to be single-exit.  So I usually do something like this:

```
  HANDLE_FUNCTION_ENTER();
  MonoWhateverHandle result = MONO_HANDLE_NEW (MonoWhatever, NULL);

  ... code code code ...
  if (!is_ok (error))
    goto leave;
  ... more code, maybe MONO_HANDLE_ASSIGN (result, successeful value) ...

leave:
  HANDLE_FUNCTION_RETURN_REF (result);
}
```